5 Ansible Deep Dive

dahachm opened the pull request:
@SagurovA93 

  Ansible playbook that:
      - create new user, add it to sudoers and deploy SSH key, so it's available to connect via SSH from one host to another (as new user)
      - install nginx and deploy a web site
      - install ntp
      - instal MySQL 5.6, create a user and a database. Here  I Use MySQL 5.6, because it doesn't store root password in log file and leave it empty (MySQL 5.7 and MySQL 8.0 do the opposite)

  5 roles were used:
      - common - to add epel-repo and update all the packages
      - user
      - ntp
      - mysql
      - nginx

dahachm committed:
Add files for task 1-3

        Add hosts.yml - inventory file
        Add set-ssh.yml - playbook file for Task 2
        Add ntp.conf - custom conf file for ntp
	Add playbook.yml - playbool file for Task 3

dahachm committed:
Create README.md

dahachm committed:
Update playbook.yml

	Add task to add MySQL 5.6 repo

dahachm committed:
Add task 4* (setup nginx web server)

	setup-nginx.yml - ansible playbook that setup web server sources and configuration
		copy web/ dir wirh recuired sources to hosts
		copy and replace nginx.conf file
		restart nginx
		add firewall rule for port 8080/tcp
		set required SELinux context to web/ dir

dahachm committed:
Update README.md

Add task 4* description

Update task 4 description ( added some sentences about used MySQL version)

dahachm committed:
Update README.md

dahachm committed:
Add hometask 5 Ansivle Deep Dive

	Task: to create an ansible playbook that performs following tasks using roles:
		- create new user on remote hosts and setup SSH keys for them
		- install ntp
		- install nginx, configure web-server and setup web-site
		- instal MySQL, create a user and db

dahachm committed:
Fix in ntp role

	Bug: ntp not starts only whe ntp.conf is updating

	Fix: Check if ntp is started in ntp/tasks every time when play ntp role

dahachm committed:
Update nginx role

	Replase nginx.conf and index.html from files/ to templates/.

	Update tasks/main.yml and hadnlers/main.yml to Copy these
	files to remote hosts using `template` module.

dahachm committed:
Update nginx role

	Add firewalld restart after adding a permanent rule to apply it

dahachm committed:
Create README.md

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/user/files/host-2_rsa:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Never ever put your private keys into the repository in plain text! 
You may use ansible-vault.

dahachm replied:
Yeah, my terrible mistake! At the begginnig I didn't plan to load keys into github repo, but I forgot to remove key files from the commit >_<

To fix this I encrypted all 3 files with Private SSH keys and updated `roles/user/tasks/main.yml`, so now:

1. File with private key and `authorized_keys` file will be updated for user (even if it's not newly created), when its content is different from the file that is already exited on remote host

2. Files with private keys will be decrypted by Ansible-vault while transfering to remote host (`--ask-vault-pass` parameter need to be used in starting playbook command)

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/user/files/host-3_rsa:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Again! Do not put private keys, passwords, etc to the git in plain text!

SagurovA93 commented on 4-Ansible_Intro/nginx.conf:
```diff
@@ -0,0 +1,70 @@
+
+    server {
+        listen       80 default_server;
+        listen       [::]:80 default_server;
```
What does it mean?

dahachm replied:
I used original `nginx.conf` file and just added new server configuration block to open it on port 8080 (the block is down below this lines). 

So default nginx website is still accesible on 80 port (for IPv4 and IPv6 on localhost)

SagurovA93 replied:
I mean `[::]:80` this string :)

SagurovA93 commented on 4-Ansible_Intro/playbook.yml:
```diff
@@ -0,0 +1,61 @@
+        name: nginx
+        state: started      
+
+  - name: Start mysql
```
This playbook will start the MySQL server even if it should not. How to deal with it? How to start MySQL when it was configured in a previous task?

SagurovA93 commented on 4-Ansible_Intro/set-ssh.yml:
```diff
@@ -0,0 +1,97 @@
+  hosts: moomin_hosts
+  tasks:
+    - name: add PubKey of host-1
+      lineinfile:
```
Could you please use [this module](https://docs.ansible.com/ansible/latest/collections/ansible/posix/authorized_key_module.html) to put ssh public keys to the remote host?

dahachm replied:
yep, already did in the next solution on the 5th branch (this is from the 4th)

SagurovA93 replied:
Ok, got it.

SagurovA93 commented on 4-Ansible_Intro/set-ssh.yml:
```diff
@@ -0,0 +1,97 @@
+          state: present
+
+    - name: add PubKey of host-2
+      lineinfile:
```
The same as in line 55

SagurovA93 commented on 4-Ansible_Intro/set-ssh.yml:
```diff
@@ -0,0 +1,97 @@
+          state: present
+
+    - name: add PubKey of host-3
+      lineinfile:
```
Please refer to the previous discussion

SagurovA93 commented on 4-Ansible_Intro/setup-nginx.yml:
```diff
@@ -0,0 +1,77 @@
+        name: nginx
+        state: present
+
+    - name: Stop nginx
+      service:
+        name: nginx
+        state: stopped
+
+    - name: Start nginx
+      service:
+        name: nginx
+        state: started          
```
You could use the `state: restarted` instead of 2 this tasks

SagurovA93 commented on 4-Ansible_Intro/setup-nginx.yml:
```diff
@@ -0,0 +1,77 @@
+        permanent: yes
+        state: enabled  
+    
+    - name: Setup SELinux context for web/
```
Could it be done with [selinux ansible](https://docs.ansible.com/ansible/latest/collections/ansible/posix/selinux_module.html) module?

SagurovA93 commented on 4-Ansible_Intro/set-ssh.yml:
```diff
@@ -0,0 +1,97 @@
+- name: Set SSH keys to host-1
+  hosts: host-1
+  
+  tasks:
+    - name: Copy ssh private key to host-1
+      copy:
+          src: host-1_rsa
+          dest: /home/moomin/.ssh/id_rsa
+          owner: moomin
+          group: moomin
+          mode: 0600
+
+- name: Copy SSH key to host-2
+  hosts: host-2
+  
+  tasks:
+    - name: Copy ssh private key to host-2
+      copy:
+          src: host-2_rsa
+          dest: /home/moomin/.ssh/id_rsa
+          owner: moomin
+          group: moomin
+          mode: 0600
```
Why you need these private keys on remote hosts?

dahachm replied:
idk, the task was to "setup SSH keys"

After adding this keys to remote hosts it's now available to login from one host to another passwordless, since we didn't set the password while creating new user.

SagurovA93 commented on 4-Ansible_Intro/set-ssh.yml:
```diff
@@ -0,0 +1,97 @@
+- name: Copy SSH key to host-3
+  hosts: host-3
+  
+  tasks:
+    - name: Copy ssh private key to host-3
+      copy:
+          src: host-3_rsa
+          dest: /home/moomin/.ssh/id_rsa
+          owner: moomin
+          group: moomin
+          mode: 0600   
```
The same as previous discussion.

SagurovA93 commented on 5-Ansible_Deep_Dive/host_vars/host-1.yml:
```diff
@@ -0,0 +1,2 @@
+private_key: files/host-1_rsa
+port_num: 8080
```
The `port_num` variable could be moved to `group_vars` because the value is the same for all hosts. On the other hand - you can define this variable only in `role/defaults.yml` if this is a default port for nginx.

dahachm replied:
Well, the idea was to add an opportunity to set custom port on each host if required, so to add this var to `group_vars` wasn't an option 

I left the same name for `port_num` var on each host to be able to change the port number using --extra-vars in command line for all the ports at the same time

SagurovA93 commented on 5-Ansible_Deep_Dive/big_installation.yml:
```diff
@@ -0,0 +1,10 @@
+- name: Setup user configuration, Install ntp, nginx and mysql
+  hosts: moomin_hosts
+
+  roles:
+        - role: user
+        - role: common
+        - role: ntp
+        - role: mysql
+        - role: nginx
```
I think this is too many spaces or tabs for this indentation. 

Usually, used 2 or 4 spaces, or 1 tab which is equal to 4 spaces. 

This is not a mistake :)

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/user/files/host-1_rsa:
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Please, put it into the vault.

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/mysql/.travis.yml:
```diff
@@ -0,0 +1,29 @@
+
+script:
+  # Basic role syntax check
+  - ansible-playbook tests/test.yml -i tests/inventory --syntax-check
```
Nice to have: run playbook in `--check` mode

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/nginx/vars/main.yml:
```diff
@@ -0,0 +1,3 @@
+# vars file for roles/nginx
+
+port_num: 8080
```
Why this variable defined in host_vars, defaults, and here?

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/user/handlers/main.yml:
```diff
@@ -0,0 +1,33 @@
+# handlers file for roles/user
+
+- name: Add new user to sudoers
+  lineinfile:
```
Are you sure this should be a handler?

dahachm replied:
The idea was to play in only if new user is created

SagurovA93 replied:
Ok

SagurovA93 commented on 5-Ansible_Deep_Dive/roles/user/handlers/main.yml:
```diff
@@ -0,0 +1,33 @@
+        line: "{{ new_user }} ALL=(ALL) NOPASSWD:ALL"
+        state: present
+
+- name: Set SSH
```
The `sshd` daemon is not responsible for user ssh keys, so it should not be started or restarted after putting the ssh keys to the remote host.

SagurovA93 reviewed (commented):
What I have to say:
1. The images for host{1,3} - amazing. 
2. Do not put private keys into the repository in plain text.

This is a good job, but please answer the questions in discussions before merging this PR.

dahachm committed:
Update roles/user

	Encrypted RSA keys with Ansible-vault and add
	using ansible-vault to decrypt them while transfering
	to remote hosts.

	Move actions with SSH setup from hadnlers to task, so SSH keys files
	will be updated (if needed) even if user is alreade created.
